Add option to keep Files in Rollback

osrecio opened the pull request:
Add option to keep files in rollback and improve Readability and PSR of Backup Library
### Description
Ask to user when launch command: `bin/magento setup:rollback` if he wants keep files to future rollbacks. Also I changed variable name protected with underscore to acomplish PSR and I imported all classes to use in class.

### Fixed Issues (if relevant)
1. magento/magento2#6460: [2.1.1 CE] Rollback/Restore deletes database (--db) backup file in ${webroot}/var/backups.

### Manual testing scenarios
1. create backup  e.g: `bin/magento setup:backup --media`
2. rollback backup e.g: `bin/magento setup:rollback -m 1508950479_filesystem_media.tgz`
3. Answer `y` to question: `you want to keep the backups?[y/N]`
4. Check that your backup was kept in `var/backups`

### Contribution checklist
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

osrecio committed:
Add option to keep files in rollback and improve Readability and use of Backup Library

fooman commented on lib/internal/Magento/Framework/Backup/Db.php:
```diff
@@ -38,14 +42,16 @@ public function rollback()
             $this->getResourceModel()->runCommand($statement);
         }
-        @unlink($source);
+        if ($this->keepSourceFile()) {
```
The logic seems to be reversed here and should be !$this->keepSourceFile()

osrecio replied:
I will check tomorrow, maybe I was wrong copying between branches

osrecio replied:
Thanks!

osrecio commented:
Hi @fooman , I've fixed these problems:
1. The problem with `lib/internal/Magento/Framework/Backup/Db.php`, I changed condition to `if (!$this->keepSourceFile()) {`
2. Problem with tests
3. Problem with codeStyle

Same for #11742 and #11737

osrecio commented:
@fooman I want to make some changes that I spoke with @dmanners  during the #MM17ES and that are already applied in the PR: #11742 .

I hope today applied for this Backport.

dmanners commented on lib/internal/Magento/Framework/Backup/AbstractBackup.php:
```diff
@@ -69,6 +72,13 @@
+     *
+     * @var boolean
+     */
+    protected $keepSourceFile;
```
Let's make this private.

osrecio replied:
Made Private

dmanners commented on lib/internal/Magento/Framework/Backup/AbstractBackup.php:
```diff
@@ -296,4 +306,21 @@ protected function _filterName($name)
+     * @param boolean $keepSourceFile
+     * @return $this
+     */
+    public function setKeepSourceFile($keepSourceFile)
```
Let's add a new interface for these methods.

osrecio replied:
Added to `SourceFileInterface`

dmanners commented on lib/internal/Magento/Framework/Backup/BackupInterface.php:
```diff
@@ -68,4 +68,19 @@ public function getType();
+     *
+     * @return boolean
+     */
+    public function keepSourceFile();
```
Let's move these to a new interface.

osrecio replied:
Added to `SourceFileInterface`

dmanners commented on setup/src/Magento/Setup/Console/Command/RollbackCommand.php:
```diff
@@ -127,7 +127,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
             }
-            $this->doRollback($input, $output);
+            $questionKeep = new ConfirmationQuestion(
+                '<info>you want to keep the backups?[y/N]<info>',
```
How about "Do you want to keep the backup files?" as the message here.

osrecio replied:
Fixed typo

osrecio commented:
@dmanners Added changes, I hope everything are Ok ;-)

dmanners committed:
Update code rollback call to pass in the rollback type of filesystem
